homeassistant/auth_providers/homeassistant.py

```diff
@@ -142,34 +262,63 @@ class HassAuthProvider(auth.AuthProvider):
             'username': username
         })
 
+    async def async_validate_2fa(self, session_token, code):
+        """Helper to validate a two factor authentication code."""
+        data = Data(self.hass, self.config.get(CONF_ENABLE_2FA))
+        await data.async_load()
+        try:
+            await self.hass.async_add_executor_job(
+                data.validate_2fa, session_token, code)
+            username = data.close_session(session_token)
+            if username is not None:
+                return username
+            raise InvalidAuth
+        finally:
+            await data.async_save()
+
 
 class LoginFlow(data_entry_flow.FlowHandler):
     """Handler for the login flow."""
 
     def __init__(self, auth_provider):
         """Initialize the login flow."""
         self._auth_provider = auth_provider
+        self._session_token = None
+        self._user_name = None
 
     async def async_step_init(self, user_input=None):
         """Handle the step of the form."""
         errors = {}
+        result = None
 
         if user_input is not None:
             try:
-                await self._auth_provider.async_validate_login(
-                    user_input['username'], user_input['password'])
+                if self._session_token and 'code' in user_input:
+                    username = await self._auth_provider.async_validate_2fa(
+                        self._session_token, user_input['code'])
+                    result = {'username': username}
+                else:
+                    await self._auth_provider.async_validate_login(
+                        user_input['username'], user_input['password'])
+                    result = {'username': user_input['username']}
             except InvalidAuth:
                 errors['base'] = 'invalid_auth'
+            except Request2FA as request2fa:
+                self._session_token = request2fa.session_token
+                errors['base'] = 'request_2fa'
 
-            if not errors:
+            if not errors and result:
                 return self.async_create_entry(
                     title=self._auth_provider.name,
-                    data=user_input
+                    data=result
                 )
 
         schema = OrderedDict()
-        schema['username'] = str
-        schema['password'] = str
+        if self._session_token:
```

Comment:
This is not how config flows work. The second step should be implemented as a second step in the flow. Example: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/hue/config_flow.py#L59